src/llama.cpp

```diff
@@ -12825,6 +12909,215 @@ struct llm_build_context {
         return gf;
     }
 
+    struct ggml_cgraph * build_minicpm3() {
```

Comment:
Is this graph very similar to `build_deepseek2()`? Can we do some code de-duplication?

Reply:
Doing a diff between this and `build_deepseek2()`, there is a lot that is in common between these two.

Some notable differences that I'm seeing on a first pass:

`minicpm3` supports scaling at several stages -- scaling the input embeddings, scaling the hidden states near the end of each layer, and finally near the end just prior to the output. This scaling factor could be set to 1 for `deepseek2` and/or skipped, and we could probably reach parity this way.

`deepseek2` supports a "lite" mode that simplifies the q calculation in each layer by a decent bit. This option just would be disabled in the `minicpm3` branch.

`deepseek2` supports MoE / shared expert calculations for generating `ffn_out`, and -- like "lite" mode -- this wouldn't be needed in the `minicpm3` branch.

`deepseek2` does some prescaling on the kq_scale and attn_factor that `minicpm3` doesn't need to do. Not sure if this could be aligned or not -- will need to dig into #7416 more before I understand this.

In calls to `ggml_rope_ext`, `minicpm3` uses rope factors that `deepseek2` simply sets to null. This could be aligned pretty easily.

And finally, the final output is calculated differently in each network -- `deepseek2` gets the result output with a call to `ggml_mul_mat`, and `minicpm3` uses a call to `llm_build_lora_mm` -- that may be one of the largest structural changes that I saw when comparing the two.

I think these two implementations could be aligned, but it may take a bit of refactoring of the deepseek2 code as well. Not sure how to weigh the value of this effort vs. just maintaining two separate branches of code.

Reply:
My view is that the best place to look for code de-duplication is in the generic 
`llm_build_xxx` and other similar functions that represent the basic building blocks of the models. Conditional branches in the implementation depending on the arch makes the code harder to follow and may have a higher maintenance cost than some code duplication.

Additionally, I think it would be preferable to create an abstract interface for the models, and move the implementation of each one to a separate file without any coupling between models, but that will be harder to implement if the different models share the same code in this way.

Reply:
> And finally, the final output is calculated differently in each network -- `deepseek2` gets the result output with a call to `ggml_mul_mat`, and `minicpm3` uses a call to `llm_build_lora_mm` -- that may be one of the largest structural changes that I saw when comparing the two.

That's probably a bug that will make loras that modify this tensors not work properly. `llm_build_lora_mm` is the right function to use when performing matrix multiplications with the weights.